### ISSUE TITLE
XSPP warns of bad XS code in Dynaloader

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12287]
Migrated from [rt.perl.org#114198](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=114198) (status was 'resolved')

Searchable as RT114198$

**Comment by p5pRT** [comment]
### From @bulk88

#### Created by @bulk88



This bug ticket is 100% cosmetic\. There are no runtime issues with it\. While compiling perl 5\.17 I noticed some warnings in the console\. DynaLoader\.xs is actually dl\_win32\.xs for me but I think this problem is in every dl\_\* file\, I traced the XS code back to Paul Marquess in 1994\, http&#8203;://perl5\.git\.perl\.org/perl\.git/blob/a0d0e21ea6ea90a22318550944fe6cb09ae10cda&#8203;:/ext/DynaLoader/dl\_next\.xs \. Day 1 of DynaLoader\. "dl\.c" the predecessor to DynaLoader had a very different api than DynaLoader\.

\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
        \.\.\\\.\.\\miniperl\.exe "\-I\.\.\\\.\.\\lib" "\-I\.\.\\\.\.\\lib" \.\.\\\.\.\\lib\\ExtUtils\\xsubpp
 \-noprototypes \-typemap \.\.\\\.\.\\lib\\ExtUtils\\typemap  DynaLoader\.xs > DynaLoader\.x
sc && C&#8203;:\\p517\\perl\\miniperl\.exe "\-I\.\.\\\.\.\\lib" "\-I\.\.\\\.\.\\lib" \-MExtUtils&#8203;::Command
\-e mv \-\- DynaLoader\.xsc DynaLoader\.c
Warning&#8203;: Found a 'CODE' section which seems to be using 'RETVAL' but no 'OUTPUT'
 section\. in DynaLoader\.xs\, line 134
Warning&#8203;: Found a 'CODE' section which seems to be using 'RETVAL' but no 'OUTPUT'
 section\. in DynaLoader\.xs\, line 162
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
I looked at the 2 lines\, first is dl\_load\_file\, which declares a void \* return type through XSPP XS syntax\, yet manually uses ST\(0\)\. Second is dl\_find\_symbol which again declares a void \* return type\, yet manually uses ST\(0\)\. Either use the XS return type the correct way with an OUTPUT section\, or don't use the XS return type feature and use ST\(\*\) and XSRETURN/XSRETURN\_\*/PPCODE and PUSH\. This ticket is 100% cosmetic\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low
    module=DynaLoader

Site configuration information for perl 5.17.2:

Configured by Owner at Mon Jul 16 14:34:12 2012.

Summary of my perl5 (revision 5 version 17 subversion 2) configuration:
   
  Platform:
    osname=MSWin32, osvers=5.1, archname=MSWin32-x86-multi-thread
    uname=''
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cl', ccflags ='-nologo -GF -W3 -Od -MD -Zi -DDEBUGGING -DWIN32 -D_CONSOLE -DNO_STRICT  -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -D_USE_32BIT_TIME_T',
    optimize='-Od -MD -Zi -DDEBUGGING',
    cppflags='-DWIN32'
    ccversion='13.10.6030', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=8
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='__int64', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -debug  -libpath:"c:\perl517\lib\CORE"  -machine:x86'
    libpth="C:\Program Files\Microsoft Visual Studio .NET 2003\VC7\lib"
    libs=oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib comctl32.lib msvcrt.lib
    perllibs=oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib comctl32.lib msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=true, libperl=perl517.lib
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -debug  -libpath:"c:\perl517\lib\CORE"  -machine:x86'

Locally applied patches:
    


@INC for perl 5.17.2:
    C:/perl517/site/lib
    C:/perl517/lib
    .


Environment for perl 5.17.2:
    HOME (unset)
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=C:\perl517\bin;C:\Program Files\Microsoft Visual Studio .NET 2003\Common7\IDE;C:\Program Files\Microsoft Visual Studio .NET 2003\VC7\BIN;C:\Program Files\Microsoft Visual Studio .NET 2003\Common7\Tools;C:\Program Files\Microsoft Visual Studio .NET 2003\Common7\Tools\bin\prerelease;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\system32\wbem;
    PERL_BADLANG (unset)
    PERL_JSON_BACKEND=JSON::PP
    SHELL (unset)

 		 	   		  
```
</details>

**Comment by p5pRT** [comment]
### From mail@steffen-mueller.net

On 07/19/2012 10&#8203;:58 AM\, bulk 88 \(via RT\) wrote&#8203;:

> This bug ticket is 100% cosmetic\. There are no runtime issues with
> it\. While compiling perl 5\.17 I noticed some warnings in the console\.
> DynaLoader\.xs is actually dl\_win32\.xs for me but I think this problem
> is in every dl\_\* file\, I traced the XS code back to Paul Marquess in
> 1994\,
> http&#8203;://perl5\.git\.perl\.org/perl\.git/blob/a0d0e21ea6ea90a22318550944fe6cb09ae10cda&#8203;:/ext/DynaLoader/dl\_next\.xs
> \. Day 1 of DynaLoader\. "dl\.c" the predecessor to DynaLoader had a
> very different api than DynaLoader\.

> I looked at the 2 lines\, first is dl\_load\_file\, which declares a
> void \* return type through XSPP XS syntax\, yet manually uses ST\(0\)\.
 > Second is dl\_find\_symbol which again declares a void \* return type\,
 > yet manually uses ST\(0\)\. Either use the XS return type the correct way
> with an OUTPUT section\, or don't use the XS return type feature and
> use ST\(\*\) and XSRETURN/XSRETURN\_\*/PPCODE and PUSH\. This ticket is
> 100% cosmetic\.

It was I who added that warning about missing OUTPUT sections\. This was 
because it is all too easy to forget to add "OUTPUT&#8203;: RETVAL" to any one 
of the gazillions XSUBs that use typemaps and people \(including me\) get 
it wrong all the time\. As far as the XS documentation goes\, RETVAL and 
an XS return type have no point without the OUTPUT section at all\. The 
warning seemed a very useful diagnostic\. In fact\, I was going to attempt 
going one step further by making an "OUTPUT&#8203;: RETVAL" implicit if RETVAL 
is used / there is a non\-void XSUB return type\. Didn't get to try it yet 
since it requires a CPAN smoke\.

The code in DynaLoader IMO works incidentially\, but relies on an XS 
implementation detail \-\- at least as far as the documented parts of XS 
are concerned\. That's not too bad since it's core\, but if we're going to 
keep the warning \(which I still consider valuable\)\, I think it's 
worthwhile to fix the warnings\. Alas\, it's per\-architecture code\, so 
mistakes will be tough to find for the implementer\. :\(

Best regards\,
Steffen

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From mail@steffen-mueller.net

On 07/19/2012 10&#8203;:58 AM\, bulk 88 \(via RT\) wrote&#8203;:
> \# New Ticket Created by  bulk 88 \# Please include the string&#8203;:  \[perl
> \#114198\] \# in the subject line of all future correspondence about
> this issue\. \# \<URL&#8203;:
> https://rt-archive.perl.org/perl5/Ticket/Display.html?id=114198 >

>
> This bug ticket is 100% cosmetic\. There are no runtime issues with
> it\. While compiling perl 5\.17 I noticed some warnings in the console\.
> DynaLoader\.xs is actually dl\_win32\.xs for me but I think this problem
> is in every dl\_\* file\, I traced the XS code back to Paul Marquess in
> 1994\,
> http&#8203;://perl5\.git\.perl\.org/perl\.git/blob/a0d0e21ea6ea90a22318550944fe6cb09ae10cda&#8203;:/ext/DynaLoader/dl\_next\.xs
> \. Day 1 of DynaLoader\. "dl\.c" the predecessor to DynaLoader had a
> very different api than DynaLoader\.

I have a fix for this \(editing the XS in DynaLoader\)\. Alas\, I cannot 
test it entirely since not all of the XS is compiled on all systems\.

Should I push this \(being cosmetic and all\)?
Risk of failure is moderately low and if it fails\, it's a dumb mistype 
easily corrected\.

\-\-Steffen

**Comment by p5pRT** [comment]
### From mail@steffen-mueller.net

On 08/13/2012 10&#8203;:14 AM\, Steffen Mueller wrote&#8203;:
> On 07/19/2012 10&#8203;:58 AM\, bulk 88 \(via RT\) wrote&#8203;:
>> \# New Ticket Created by  bulk 88 \# Please include the string&#8203;:  \[perl
>> \#114198\] \# in the subject line of all future correspondence about
>> this issue\. \# \<URL&#8203;:
>> https://rt-archive.perl.org/perl5/Ticket/Display.html?id=114198 >

>> This bug ticket is 100% cosmetic\. There are no runtime issues with
>> it\. While compiling perl 5\.17 I noticed some warnings in the console\.
>> DynaLoader\.xs is actually dl\_win32\.xs for me but I think this problem
>> is in every dl\_\* file\, I traced the XS code back to Paul Marquess in
>> 1994\,

> I have a fix for this \(editing the XS in DynaLoader\)\. Alas\, I cannot
> test it entirely since not all of the XS is compiled on all systems\.
>
> Should I push this \(being cosmetic and all\)?
> Risk of failure is moderately low and if it fails\, it's a dumb mistype
> easily corrected\.

Pushed as smoke\-me branch at
smoke\-me/dynaloader\_silence\_xs\_warning

Best regards\,
Steffen

**Comment by p5pRT** [comment]
### From @nwc10

On Mon\, Aug 13\, 2012 at 10&#8203;:32&#8203;:47AM \+0200\, Steffen Mueller wrote&#8203;:
> On 08/13/2012 10&#8203;:14 AM\, Steffen Mueller wrote&#8203;:

> > Should I push this \(being cosmetic and all\)?
> > Risk of failure is moderately low and if it fails\, it's a dumb mistype
> > easily corrected\.
> 
> Pushed as smoke\-me branch at
> smoke\-me/dynaloader\_silence\_xs\_warning

Please don't merge this yet\. It's broken AIX somehow\, but I don't have
time yet to work out why or fix it\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @tsee

On 08/17/2012 06&#8203;:31 PM\, Nicholas Clark wrote&#8203;:
> On Mon\, Aug 13\, 2012 at 10&#8203;:32&#8203;:47AM \+0200\, Steffen Mueller wrote&#8203;:
>> On 08/13/2012 10&#8203;:14 AM\, Steffen Mueller wrote&#8203;:
>
>>> Should I push this \(being cosmetic and all\)?
>>> Risk of failure is moderately low and if it fails\, it's a dumb mistype
>>> easily corrected\.
>>
>> Pushed as smoke\-me branch at
>> smoke\-me/dynaloader\_silence\_xs\_warning
>
> Please don't merge this yet\. It's broken AIX somehow\, but I don't have
> time yet to work out why or fix it\.

Thanks for the heads\-up\. I'd bet it's a dumb typing fail on my part\. 
I'll try to review the code again\, maybe something obvious jumps out\.

\-\-Steffen

**Comment by p5pRT** [comment]
### From @bulk88

On Fri Aug 17 09&#8203;:32&#8203;:23 2012\, nicholas wrote&#8203;:
> Please don't merge this yet\. It's broken AIX somehow\, but I don't have
> time yet to work out why or fix it\.
> 
> Nicholas Clark
> 
Do you have any build logs or smoke reports for the AIX breakage?

**Comment by p5pRT** [comment]
### From @bulk88

On Fri Aug 17 10&#8203;:18&#8203;:45 2012\, smueller@&#8203;cpan\.org wrote&#8203;:
> Thanks for the heads\-up\. I'd bet it's a dumb typing fail on my part\. 
> I'll try to review the code again\, maybe something obvious jumps out\.
> 
> \-\-Steffen
>
from dl\_aix\.xs
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
\-void \*
\+void
 dl\_load\_file\(filename\, flags=0\)
&nbsp;        char \*  filename
&nbsp;        int     flags
\-       CODE&#8203;:
\+        PREINIT&#8203;:
\+        void \*retv;
\+       PPCODE&#8203;:
&nbsp;        DLDEBUG\(1\,PerlIO\_printf\(Perl\_debug\_log\,
"dl\_load\_file\(%s\,%x\)&#8203;:\\n"\, filename\,flags\)\);
&nbsp;        if \(flags & 0x01\)
&nbsp;            Perl\_warn\(aTHX\_ "Can't make loaded symbols global on this
platform while loading %s"\,filename\);
\-       RETVAL = dlopen\(filename\, RTLD\_GLOBAL|RTLD\_LAZY\) ;
\-       DLDEBUG\(2\,PerlIO\_printf\(Perl\_debug\_log\, " libref=%x\\n"\, RETVAL\)\);
\+       retv = dlopen\(filename\, RTLD\_GLOBAL|RTLD\_LAZY\) ;
\+       DLDEBUG\(2\,PerlIO\_printf\(Perl\_debug\_log\, " libref=%x\\n"\, retv\)\);
&nbsp;        ST\(0\) = sv\_newmortal\(\) ;
\-       if \(RETVAL == NULL\)
\+       if \(retv == NULL\)
&nbsp;            SaveError\(aTHX\_ "%s"\,dlerror\(\)\) ;
&nbsp;        else
\-           sv\_setiv\( ST\(0\)\, PTR2IV\(RETVAL\) \);
\+           sv\_setiv\( ST\(0\)\, PTR2IV\(retv\) \);
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
If you do a PPCODE\, 0 items will be returned by the secret PUTBACK\. You
need PUSH\* or XSRETURN\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Aug 17 13&#8203;:02&#8203;:06 2012\, bulk88 wrote&#8203;:
> If you do a PPCODE\, 0 items will be returned by the secret PUTBACK\. You
> need PUSH\* or XSRETURN\.

That’s bitten me before\, too\.  The whole point of PPCODE is to allow the
XS author to do the stack manipulation\.  But then xsubpp does part of
it\.  It’s probably too late to change it now\, isn’t it?

\(BTW\, ext/File\-Glob/Glob\.xs has a dummy SPAGAIN at the end of a function
to work around this\.\)

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @nwc10

On Fri\, Aug 17\, 2012 at 01&#8203;:02&#8203;:07PM \-0700\, bulk 88 via RT wrote&#8203;:
> On Fri Aug 17 10&#8203;:18&#8203;:45 2012\, smueller@&#8203;cpan\.org wrote&#8203;:
> > Thanks for the heads\-up\. I'd bet it's a dumb typing fail on my part\. 
> > I'll try to review the code again\, maybe something obvious jumps out\.
> > 
> > \-\-Steffen
> >
> from dl\_aix\.xs

> If you do a PPCODE\, 0 items will be returned by the secret PUTBACK\. You
> need PUSH\* or XSRETURN\.

Which means that this hack isn't the right answer&#8203;:

<details><summary>Inline Patch</summary>

```diff
diff --git a/ext/DynaLoader/dl_aix.xs b/ext/DynaLoader/dl_aix.xs
index 07730e9..a74d23b 100644
--- a/ext/DynaLoader/dl_aix.xs
+++ b/ext/DynaLoader/dl_aix.xs
@@ -694,7 +694,7 @@ dl_load_file(filename, flags=0)
        int     flags
         PREINIT:
         void *retv;
-       PPCODE:
+       CODE:
        DLDEBUG(1,PerlIO_printf(Perl_debug_log, "dl_load_file(%s,%x):\n", filena

```
</details>
me,flags));
&nbsp;        if \(flags & 0x01\)
&nbsp;            Perl\_warn\(aTHX\_ "Can't make loaded symbols global on this platform w
hile loading %s"\,filename\);

but it does get this working again&#8203;:

$ \./perl \-Ilib \-MFcntl \-e0
$

\(build still running\, tests then follow\, but I'll be asleep\)

On Fri\, Aug 17\, 2012 at 12&#8203;:27&#8203;:46PM \-0700\, bulk 88 via RT wrote&#8203;:
> On Fri Aug 17 09&#8203;:32&#8203;:23 2012\, nicholas wrote&#8203;:
> > Please don't merge this yet\. It's broken AIX somehow\, but I don't have
> > time yet to work out why or fix it\.
> > 
> > Nicholas Clark
> > 
> Do you have any build logs or smoke reports for the AIX breakage?

No\, sorry\. It's an interactive login\, running manually\, so there's nothing
automatic\. I can run things\, but you already seem to have hit the nail on the
head as to the cause\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @tsee

On 08/17/2012 11&#8203;:29 PM\, Nicholas Clark wrote&#8203;:
> On Fri\, Aug 17\, 2012 at 01&#8203;:02&#8203;:07PM \-0700\, bulk 88 via RT wrote&#8203;:
>> On Fri Aug 17 10&#8203;:18&#8203;:45 2012\, smueller@&#8203;cpan\.org wrote&#8203;:
>>> Thanks for the heads\-up\. I'd bet it's a dumb typing fail on my part\.
>>> I'll try to review the code again\, maybe something obvious jumps out\.
>>>
>>> \-\-Steffen
>>>
>> from dl\_aix\.xs
>
>> If you do a PPCODE\, 0 items will be returned by the secret PUTBACK\. You
>> need PUSH\* or XSRETURN\.
>
> Which means that this hack isn't the right answer&#8203;:

\[snip\]

But I guess simply adding the correct XSRETURN\(1\) is? Could you try 
building on AIX again using the new branch

&nbsp;   smoke\-me/dynaloader\_silence\_xs\_warning3

?

Best regards\,
Steffen

**Comment by p5pRT** [comment]
### From @nwc10

On Thu\, Aug 30\, 2012 at 04&#8203;:12&#8203;:05PM \+0200\, Steffen Mueller wrote&#8203;:

> But I guess simply adding the correct XSRETURN\(1\) is? Could you try 
> building on AIX again using the new branch
> 
>    smoke\-me/dynaloader\_silence\_xs\_warning3

Builds\, and with no errors or warnings\.

\(And all tests pass\, apart from the usual t/run/cloexec\.t\,
which in parallel "Failed tests&#8203;:  18\-20\, 22"\, but runs cleanly when I run it
on its own\)

As far as AIX is concerned\, this seems good to go\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @steve-m-hay

Fix was applied in commit 63d7ac5fa5ef436afa9865a56b7c84e54a112722\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Fix was applied in commit 63d7ac5fa5ef436afa9865a56b7c84e54a112722\.

**Comment by p5pRT** [comment]
@steve-m-hay - Status changed from 'open' to 'resolved'
